File: acmicpc.net/7576/7576_outOfMem.cpp
#include <iostream>
#include <vector>
#include <deque>

using namespace std;

int main(){
    int M, N, tmp;
    cin >> M >> N;

    vector< vector<int> > box(N);
    vector< vector<int> > nextBox(N);
    deque<int> points;
    deque<int> nextPoints;
    for (int row = 0; row < N; row++) {
        box[row] = vector<int>(M);
        nextBox[row] = vector<int>(M);
        for (int col = 0; col < M; col++) {
            cin >> tmp;
            box[row][col] = tmp;
            nextBox[row][col] = tmp;
            if (tmp == 1){
                points.push_back(row);
                points.push_back(col);
            }
        }
    }

    int result = -1;
    while (points.size() > 0) {
        cout << "test " << points.size() << endl;
        result++;
        while (points.size() > 0){
            int row = points.front();
            points.pop_front();
            int col = points.front();
            points.pop_front();

            nextBox[row][col] = 1;
            if (col > 0 && box[row][col - 1] == 0) {
                nextBox[row][col - 1] = 1;
                nextPoints.push_back(row);
                nextPoints.push_back(col - 1);
            }
            if (col < M - 1 && box[row][col + 1] == 0) {
                nextBox[row][col + 1] = 1;
                nextPoints.push_back(row);
                nextPoints.push_back(col + 1);
            }
            if (row > 0 && box[row - 1][col] == 0) {
                nextBox[row - 1][col] = 1;
                nextPoints.push_back(row - 1);
                nextPoints.push_back(col);
            }
            if (row < N - 1 && box[row + 1][col] == 0) {
                nextBox[row + 1][col] = 1;
                nextPoints.push_back(row + 1);
                nextPoints.push_back(col);
            }
        }
        swap(box, nextBox);
        swap(points, nextPoints);
    }

    for (int row = 0; row < N; row++) {
        for (int col = 0; col < M; col++) {
            if (box[row][col] == 0) {
                cout << -1 << endl;
                return 0;
            }
        }
    }
    cout << result << endl;
    return 0;
}

File: acmicpc.net/7576/7576_outOfMem.py
from collections import deque

def check(BOX):
    for row in BOX:
        for item in row:
            if item == 0:
                return False
    return True

if __name__ == "__main__":
    M, N = map(int, input().split())

    BOX = []
    NEXTBOX = []
    queue = deque()
    nextQueue = deque()
    for i in range(N):
        BOX.append(list(map(int, input().split())))
        NEXTBOX.append(BOX[-1][:])
        for j in range(M):
            if BOX[i][j] == 1:
                queue.append((i, j))
    
    res = 0
    while queue:
        res += 1
        changed = False
        while queue:
            row, col = queue.popleft()

            NEXTBOX[row][col] = 1
            if col > 0 and BOX[row][col - 1] == 0:
                NEXTBOX[row][col - 1] = 1
                nextQueue.append((row, col - 1))
            if col <  M - 1 and BOX[row][col + 1] == 0:
                NEXTBOX[row][col + 1] = 1
                nextQueue.append((row, col + 1))
            if row > 0 and BOX[row - 1][col] == 0:
                NEXTBOX[row - 1][col] = 1
                nextQueue.append((row - 1, col))
            if row < N - 1 and BOX[row + 1][col] == 0:
                NEXTBOX[row + 1][col] = 1
                nextQueue.append((row + 1, col))
        BOX, NEXTBOX = NEXTBOX, BOX
        queue, nextQueue = nextQueue, queue
    
    if check(BOX):
        print(res - 1)
    else:
        print(-1)
